### ISSUE TITLE
A way to specify build-dependencies

### ISSUE THREAD

**Issue by con-f-use** [pypa/pipenv#3651]
Pipenv might need a way to specify built-dependencies and tooling packages that need to be installed in order for pipenv to do it's job correctly.

## Example

In my company we use a wrapper around setuptools to simplify writing `setup.py` scripts for our packaging. Let's call this wrapper `stwrapper`.

A typical `setup.py` would look like this:

```
# setup.py for "package-that-uses-stwrapper-in-its-setup-py"
from stwrapper import setup

setup(
   name='package-that-uses-stwrapper-in-its-setup-py',
   # ... some more arguments
)
```
Where the function `stwrapper.setup` parses arguments, does some extra work and then calls the real `setuptools.setup`. Effectively, the `stwrapper` package is sort of a built dependency for us.

Lets say we have the following `Pipfile`:
```
[dev-packages]
stwrapper = "*"
package-that-uses-stwrapper-in-its-setup-py = { path = ".", editable = true, setup_requires=["stwrapper"] }
```
and do a 
```
pipenv lock --verbose
```
then, with the current version of Pipenv, we get:
```
Traceback:
...
pipenv.patched.notpip._internal.exceptions.InstallationError: Command "python setup.py egg_info" failed with error code 1 in '.'
```
Because `stwarpper` is not installed at the time pipenv tries to parse the `setup.py` of `package-that-uses-stwrapper-in-its-setup-py`. That is even the case, when `stwrapper` **is installed** in the **system's python** installation. Pipenv wants to have it **inside the new virtual environment it creates**. It's a Catch-22.

Our current workaround is to let pipenv create the virutal environment. Then do a `pipenv run pip install stwrapper` inside it and then `pipenv lock`. That way it works.

Is there a solution for this already, that I'm not aware of?

## Possible Solutions

 - Pipenv gets an equivalent to `setup_requires` in setup files or installs explicitly specified `setup_requires`
 - Pipenv uses the system python's installation of `stwrapper` (I'm aware of the `--site-packages` option but that may have unwanted side effects). 
 - We can specify packages in the `requires` section of a pipfile that need to be installed before Pipenv does any work, i.e. before locking. 
    ```
    [requires]
    python_version = "2.7"
    stwrapper = "*"
    ```
## Scope

I guess the problem applies to all projects that use non-standard-libaray packages in their `setup.py` files and/or use setuptools entry points to modify its behavior. You can not put such dependencies in the Pipfile, because they are needed for successfully locking.

## Related Issues

 - https://github.com/pypa/pipenv/issues/2274

**Comment by uranusjr** [comment]
The canonical way is to fix packaging metadata in `package-that-uses-stwrapper-in-its-setup-py` (in other words, the package maintainers should be responsible to fixing this, not the packaging tools).

I believe the canonical way now is to enable build isolation (PEP 517), and leverage against [PEP 518](https://www.python.org/dev/peps/pep-0518/) (which is available in pip 19.x):

```toml
[build-system]
requires = ['setuptools', 'wheel', 'stwrapper']
```

**Comment by con-f-use** [comment]
Well but if you do a **local** install (`pip install -e .`), the package metadata is not accessible until the local `setup.py` is run. But it cannot be run, because in the virtual environment `stwrapper` is not yet isnstalled.

**Comment by uranusjr** [comment]
They are already implemented, and supported by pip and setuptools.

**Comment by con-f-use** [comment]
Okay, I just learned that the `build-system` solution, does not work in case of [editable installs](https://github.com/pypa/pip/issues/6375).

See [PEP 517](https://www.python.org/dev/peps/pep-0517/):
> This PEP [517] originally specified another hook, install_editable, to do an editable install (as with pip install -e). It was removed due to the complexity of the topic, but may be specified in a later PEP.
> 
> Briefly, the questions to be answered include: what reasonable ways existing of implementing an 'editable install'? Should the backend or the frontend pick how to make an editable install? And if the frontend does, what does it need from the backend to do so.

**Comment by con-f-use** [comment]
Passing options down to pip like this
```
[dev-packages]
mypackage = { path = ".", editable = true, pip_options=["--no-use-pep517"] } 
```
seems to have been discussed here https://github.com/pypa/pipenv/issues/287

**Comment by uranusjr** [comment]
I wonder if this can be fixed more simply by passing `--no-use-pep517` during the retry phase of installation.

**Comment by con-f-use** [comment]
Since https://github.com/pypa/pip/pull/6370 will implement a `editable mode is not supported for pyproject.toml-style projects [...] you may pass --no-use-pep517 [...]` for pip, Pipenv could look for that output and retry with `--no-use-pep517`. I so no reason not to.

**Comment by uranusjr** [comment]
@techalchemy What do you think about this, should the change require a PEEP? I’ll submit a fix if that’s not necessary.

**Comment by con-f-use** [comment]
Any work done yet?

**Comment by pawelswiecki** [comment]
What is the current status of this issue?

**Comment by con-f-use** [comment]
The more I use pipenv, the more I'd like to pass pip-options on a per-package basis (i.e. no. 4 of my [proposed solutions](https://github.com/pypa/pipenv/issues/3651#issue-425900928)).

@uranusjr @kenneth-reitz @techalchemy @nateprewitt

**Comment by kenneth-reitz** [comment]
This seems like a potential good idea.

**Comment by con-f-use** [comment]
Is there any progress here? When is the next release scheduled? At least this seems straight forward to implement:

> Since [pypa/pip#6370](https://github.com/pypa/pip/pull/6370) will implement a `editable mode is not supported for pyproject.toml-style projects [...] you may pass --no-use-pep517 [...]` for pip, Pipenv could look for that output and retry with `--no-use-pep517`. I so no reason not to.

**Comment by con-f-use** [comment]
Btw. passing options down to pip generally seems to be a good idea because otherwise, **locking fails on Windows** for packages like [pycurl](https://github.com/pycurl/pycurl), because they try to build but require some additional options (see: https://github.com/pycurl/pycurl/blob/master/setup.py#L388):

```
    def configure_windows(self):
        # Windows users have to pass --curl-dir parameter to specify path
        # to libcurl, because there is no curl-config on windows at all.
        curl_dir = scan_argv(self.argv, "--curl-dir=")
        if curl_dir is None:
            fail("Please specify --curl-dir=/path/to/built/libcurl")
```

**Comment by techalchemy** [comment]
I agree this would be a good idea, someone just needs to write up a proposal (if you've done this already just point me at it)

essentially there's no disagreement about whether this should exist, more just no clear picture of how it should look if we are going to implement it.  I don't have any particular uses for it so I would probably have a hard time designing the interface.  For instance, would it just be an additional key in the `Pipfile` entry? What would it say?

That would be outlined in a `PEEP` as a PR to the `peeps/` directory with a new number just saying the basics and following the templates that some of the other enhancements in there have followed.  Just outline the desired behavior and how it should look, plus if there are any concerns or considerations that might make this hard to implement.  Ideally, we want this to solve as general a case as possible so we don't have to reimplement this a bunch of different ways

Hope that provides a good path forward on this

**Comment by con-f-use** [comment]
As I said, I'd be happy with something like this:
```
[dev-packages]
mypackage = { path = ".", editable = true, extra_pip_options=["--no-use-pep517", "--retries=10"] } 
```
But I do not know whether it's consistent with how other things work in pipenv. Also I have no idea wow to write a PEEP properly. And there is the question of how to handle it if pipenv would pass an argument to pip and it's also specified in `extra_pip_options`. Should the extras overwrite pipenvs setting? Should it be the other way arround? Should we get an "overwriting pip option" error? Should we get an error and have the option to just "prefer extras"  or "prefere pipenv options"?

**Comment by matteius** [comment]
This is a 2019 issue about specifying build dependencies. The core problem is about packages that need other packages installed before they can even be locked/built (e.g., custom setup.py wrappers). The suggested solutions include requires=[...] per-package, pip_options=[...], or installation ordering.

However, this is 2026 now and PEP 517/518 with pyproject.toml [build-system] requires has been the standard solution for years. Editable installs with PEP 517 are also fully supported since PEP 660
